File: Baekjun/9.4/1311 할 일 정하기 1/solution.py
import sys

input = sys.stdin.readline
n = int(input())
cost = []
MAX = 100000
dp = [MAX * 21 for _ in range(1 << n)]
for _ in range(n):
    cost.append(list(map(int, input().split())))
dp[0] = 0


for i in range(1 << n):
    cnt = 0
    j = i

    while j:
        j -= j & -j
        cnt += 1

    for j in range(n):
        if i & (1 << j):
            continue

        dp[i | (1 << j)] = min(dp[i | (1 << j)], dp[i] + cost[cnt][j])
print(dp[(1 << n) - 1])

File: Baekjun/9.4/1311 할 일 정하기 1/solution.swift
let N = Int(readLine()!)!

var cost:[[Int]] = []

for _ in 0..<N {
    cost.append(readLine()!.split(separator: " ").compactMap{Int($0)})
}
 
let MAX = 10000
var dp = Array(repeating:MAX, count: (1 << N))

dp[0] = 0

for i in 0..<(1 << N) {
    let cnt = i.bitWidth
    for j in 0..<N {
        if i & (1 << j) {
            continue
        }
        dp[i | (1 << j)] = min(dp[i | (1 << j)], dp[i] + cost[cnt][j])
    }
}

print(dp.last!)